[android] Set Badge Count Number Based on Notification Payload

danstepanov opened the pull request:
# Why

The badge field set in notification payload is ignored on Android. This PR updates it so that it is passed via the notification badge field, which different launchers treat differently.

# How

Set badge count number by fetching the value from the payload content.

# Test Plan

1. Install microsoft launcher on android device and turn on badges: https://www.windowscentral.com/how-turn-notification-badges-microsoft-launcher
2. Patch PR
3. Build & Debug Expo Go in android studio on device
4. Load `native-component-list` app running locally
5. Open Notifications screen, take note of push token printed to console
6. Start www pointing at production locally, modified to pass the `badge` field through to FCM.
7. Send curl to local prod with token:
```
curl --request POST \
  --url http://localhost:3000/--/api/v2/push/send \
  --header 'Content-Type: application/json' \
  --data '{
	"to": "ExponentPushToken[...]",
	"data": {"hello": "world"},
	"badge": 10
}'
```
8. See notification is received on device and badge is incremented by specified value.

![screen](https://user-images.githubusercontent.com/189568/167515786-64250510-82e5-46c0-8d23-1f339531a092.png)


# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

MohamedAbdalkader reviewed (approved):
Thanks!

brentvatne commented on android/versioned-abis/expoview-abi44_0_0/src/main/java/abi44_0_0/expo/modules/notifications/notifications/presentation/builders/CategoryAwareNotificationBuilder.java:
```diff
@@ -36,6 +36,8 @@ protected NotificationCompat.Builder createBuilder() {
       addActionsToBuilder(builder, categoryIdentifier);
     }
+
+    builder.setNumber(content.getBadgeCount().intValue());
```
`getBadgeCount()` is nullable

danstepanov replied:
good call

bbarthec reviewed (commented):
Looking good 👍 
Please add [CHANGELOG entry](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting) and we're ready to merge it 💪

bbarthec reviewed (commented):
Just want to make sure here: you want to modify `SDK 44` codebase? 🤔 
Because you've modified versioned version of this file.
Don't you want to modify `UNVERSIONED`/`SDK 45` file?
I guess you want to modify [`packages/expo-notification/.../CategoryAwareNotificationBuilder.java`](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/presentation/builders/CategoryAwareNotificationBuilder.java) and maybe backport it to the versioned SDK 45 in the file [`android/versioned-abis/expoview-abi45_0_0/.../CategoryAwareNotificationBuilder.java`](https://github.com/expo/expo/blob/main/android/versioned-abis/expoview-abi45_0_0/src/main/java/abi45_0_0/expo/modules/notifications/notifications/presentation/builders/CategoryAwareNotificationBuilder.java)

wschurman reviewed (changes requested):
So it looks like the idea was for there to be a special payload that the user would send to change the badge: https://github.com/expo/expo/pull/6849

Then, during a refactor, the functionality was intentionally orphaned (with a plan to implement it down the road according to the summary): https://github.com/expo/expo/pull/7068

It appears that the plan to un-orphan the badge stuff never happened, and the stuff that sets the badge remains unused (`BadgeSettingNotificationBuilder`).

As for where to go from here:
- We should probably remove the orphaned stuff just to simplify this.
- We can do this in the `CategoryAwareNotificationBuilder` (like this PR does)

Before this can land though, it'll need to address a few things:
- @bbarthec's comment about versions
- A test plan

danstepanov commented:
> So it looks like the idea was for there to be a special payload that the user would send to change the badge: #6849
> 
> Then, during a refactor, the functionality was intentionally orphaned (with a plan to implement it down the road according to the summary): #7068
> 
> It appears that the plan to un-orphan the badge stuff never happened, and the stuff that sets the badge remains unused (`BadgeSettingNotificationBuilder`).
> 
> As for where to go from here:
> 
> * We should probably remove the orphaned stuff just to simplify this.
> * We can do this in the `CategoryAwareNotificationBuilder` (like this PR does)
> 
> Before this can land though, it'll need to address a few things:
> 
> * @bbarthec's comment about versions
> * A test plan

@wschurman I'm not sure what the test plan here should be, what is typically set up to test something like this?

bbarthec commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@ Package-specific changes not released in any SDK will be added here just before
 - Updated `@stripe/stripe-react-native` from `0.2.3` to `0.6.0`. ([#16938](https://github.com/expo/expo/pull/16938) by [@kudo](https://github.com/kudo))
 - Updated `react-native-reanimated` from `2.4.1` to `2.8.0`. ([#16956](https://github.com/expo/expo/pull/16956), [#17159](https://github.com/expo/expo/pull/17159) by [@tsapeta](https://github.com/tsapeta))
 - Updated `react-native-maps` from `0.29.4` to `0.30.1`. ([#16944](https://github.com/expo/expo/pull/16944) by [@bycedric](https://github.com/bycedric))
+- Set badge count number on android by fetching the value from the notification payload content. ([#17171](https://github.com/expo/expo/pull/17171) by [@danstepanov](https://github.com/danstepanov))
```
Move this entry to the `packages/expo-notifications/CHANGELOG` and place it under [`Unpublished -> Bug fixes` section](https://github.com/expo/expo/blob/main/packages/expo-notifications/CHANGELOG.md#-bug-fixes)
```suggestion
- On Android fix the `badge count number` not being read from the notification payload content. ([#17171](https://github.com/expo/expo/pull/17171) by [@danstepanov](https://github.com/danstepanov))
```

tsapeta replied:
I also recommend reading this guide [Updating Changelogs](https://github.com/expo/expo/blob/main/guides/contributing/Updating%20Changelogs.md) for more details on how we use changelogs 😉

danstepanov replied:
yea was late, forgot to move the changelog, thanks for flagging this

bbarthec reviewed (changes requested):
Looking better 👍 
Move CHANGELOG entry to the respective package.

Re: testing this feature: in `native-component-list` app there's a `Notification` screen with two actions that might affect `badge count`:
![Screenshot_20220426_110735](https://user-images.githubusercontent.com/16623003/165265084-1c9619d1-e891-4763-b6cd-e10500484efe.png)

After successful merge we'll have to make sure this change is propagated correctly for `SDK-45` if needed (publishing the package on npm, rebuilding `aar`, etc)

wschurman commented:
> Re: testing this feature: in `native-component-list` app there's a `Notification` screen with two actions that might affect `badge count`

For this PR in particular, the "Badge Number" section is separate (it's a JS-only API). We'll want to change the "Send me a push notification" test to change the badge via this new field and verify it that way.

> After successful merge we'll have to make sure this change is propagated correctly for `SDK-45` if needed (publishing the package on npm, rebuilding `aar`, etc)

I don't think we need to merge this to 45 since it's a new feature (ish) and late in the release cycle, but I could be convinced otherwise.

MohamedAbdalkader commented:
> So it looks like the idea was for there to be a special payload that the user would send to change the badge: https://github.com/expo/expo/pull/6849

@wschurman  Worth noting that this is different. The PR you referenced uses shortcut badger which would try to imitate iOS behaviour on Android devices. The change in this PR is leveraging the official android APIs that would set a badge count per notification. 

https://developer.android.com/training/notify-user/badges

danstepanov commented:
attempting to run the app on android to test the functionality lands me on this screen when running `expo start` and clicking `a` to run android within native-component-list.

![image](https://user-images.githubusercontent.com/5482800/165402454-94239abb-0534-492f-8647-93c1086c9abd.png)

Seems like y'all are able to run the app, would like to briefly pair to get this across the finish line

tsapeta commented on packages/expo-notifications/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 ### 🎉 New features
 
 ### 🐛 Bug fixes
+- On Android, fix the `badge count number` not being read from the notification payload content. ([#17171](https://github.com/expo/expo/pull/17171) by [@danstepanov](https://github.com/danstepanov))
```
Just to be consistent 😄 

```suggestion

- On Android, fix the `badge count number` not being read from the notification payload content. ([#17171](https://github.com/expo/expo/pull/17171) by [@danstepanov](https://github.com/danstepanov))
```

tsapeta commented on android/versioned-abis/expoview-abi45_0_0/src/main/java/abi45_0_0/expo/modules/notifications/notifications/presentation/builders/CategoryAwareNotificationBuilder.java:
```diff
@@ -36,6 +36,10 @@ protected NotificationCompat.Builder createBuilder() {
     if (categoryIdentifier != null) {
       addActionsToBuilder(builder, categoryIdentifier);
     }
+
+    if (content.getBadgeCount() != null) {
+      builder.setNumber(content.getBadgeCount().intValue());
+    }
```
```suggestion
```

wschurman commented:
Test steps we took:
1. Patch PR
2. Build & Debug Expo Go in android studio on device (or sim)
3. Breakpoint new code
4. Load `native-component-list` app running locally, modify the JS to also set the `badge` key in the payload on the test send push notification item. (Note: Expo Push Notifications doesn't support passing this field through on android, so the way we tested this was to add support for it in the server and then run the server locally to send a push notification rather than use the one in the `native-component-list` app).
5. Send a push containing the badge key to the app.
6. See that the breakpoint is hit.

Note that on all the launchers we tried this on, no appearance was modified. That doesn't mean it doesn't work, it just instead means that the behavior of setting this field is super nondeterministic, which makes me cautious about setting this field.
The docblock for this field mentions the inconsistent behavior: https://developer.android.com/reference/kotlin/android/app/Notification#number:kotlin.Int
On FCM, it seems that they use a different field `notification_count` to distinguish it from the legacy `badge` field in their old API. https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#androidnotification

So, going forward it may make more sense to mimic the FCM API here and use the `notification_count` field though that would be a lot more work.

danstepanov committed:
[android] Set Badge Count Number Based on Notification Payload

The badge field set in notification payload is ignored on Android. Set
badge count number by fetching the value from the payload content.

wschurman reviewed (approved):
Updated the test plan. One more change.

wschurman committed:
Update packages/expo-notifications/CHANGELOG.md

bbarthec commented on packages/expo-notifications/CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ### 🛠 Breaking changes
 
+- On Android, fix the `badge count number` not being read from the notification payload content. ([#17171](https://github.com/expo/expo/pull/17171) by [@danstepanov](https://github.com/danstepanov))
```
@wschurman, we haven't used such labeling style like `[android]` in the CHANGELOG so far.
Why do we have to change it now? Is there any specific reason backing your suggestion? 🤔 
Moreover, we tend to be very accurate in terms of proper naming (since https://github.com/expo/expo/pull/17134), so it would be `Android` instead of `android`.
